File: sdk/golang/sfa/logging.go
package sfa

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

// LogEntry is a single JSONL log entry for an agent execution.
type LogEntry struct {
	Timestamp     string         `json:"timestamp"`
	Agent         string         `json:"agent"`
	Version       string         `json:"version"`
	ExitCode      int            `json:"exitCode"`
	DurationMs    int64          `json:"durationMs"`
	Depth         int            `json:"depth"`
	CallChain     []string       `json:"callChain"`
	InputSummary  string         `json:"inputSummary"`
	OutputSummary string         `json:"outputSummary"`
	SessionID     string         `json:"sessionId"`
	Meta          map[string]any `json:"meta,omitempty"`
}

// LoggingConfig controls execution log behavior.
type LoggingConfig struct {
	FilePath     string
	Suppressed   bool
	MaxSizeBytes int64
	RetainCount  int
}

const (
	defaultMaxLogSize  = 50 * 1024 * 1024 // 50 MB
	defaultRetainCount = 5
)

// resolveLoggingConfig determines logging configuration from env, config, and flags.
func resolveLoggingConfig(config map[string]any, noLogFlag bool) *LoggingConfig {
	lc := &LoggingConfig{
		MaxSizeBytes: defaultMaxLogSize,
		RetainCount:  defaultRetainCount,
	}

	// Check suppression
	if noLogFlag || os.Getenv("SFA_NO_LOG") == "1" {
		lc.Suppressed = true
		return lc
	}

	// Resolve file path
	if p := os.Getenv("SFA_LOG_FILE"); p != "" {
		lc.FilePath = p
	} else if logging, ok := config["logging"]; ok {
		if lm, ok := logging.(map[string]any); ok {
			if f, ok := lm["file"].(string); ok {
				lc.FilePath = f
			}
			if ms, ok := lm["maxSize"].(float64); ok {
				lc.MaxSizeBytes = int64(ms) * 1024 * 1024
			}
			if rc, ok := lm["retainFiles"].(float64); ok {
				lc.RetainCount = int(rc)
			}
		}
	}

	if lc.FilePath == "" {
		home, err := os.UserHomeDir()
		if err != nil {
			lc.Suppressed = true
			return lc
		}
		lc.FilePath = filepath.Join(home, ".local", "share", "single-file-agents", "logs", "executions.jsonl")
	}

	return lc
}

// createLogEntry builds a log entry from execution data.
func createLogEntry(agent, version string, exitCode int, startTime time.Time,
	depth int, chain []string, sessionID, input, output string) *LogEntry {
	return &LogEntry{
		Timestamp:     time.Now().UTC().Format(time.RFC3339),
		Agent:         agent,
		Version:       version,
		ExitCode:      exitCode,
		DurationMs:    time.Since(startTime).Milliseconds(),
		Depth:         depth,
		CallChain:     chain,
		InputSummary:  truncate(input, 500),
		OutputSummary: truncate(output, 500),
		SessionID:     sessionID,
	}
}

// writeLogEntry appends a log entry to the log file.
// Best-effort: failures are warned to stderr but don't affect the exit code.
func writeLogEntry(entry *LogEntry, config *LoggingConfig) {
	if config.Suppressed {
		return
	}

	// Create log directory
	dir := filepath.Dir(config.FilePath)
	if err := os.MkdirAll(dir, 0755); err != nil {
		fmt.Fprintf(os.Stderr, "warning: failed to create log directory: %v\n", err)
		return
	}

	// Check if rotation is needed
	if info, err := os.Stat(config.FilePath); err == nil {
		if info.Size() >= config.MaxSizeBytes {
			rotateLog(config)
		}
	}

	// Marshal entry
	data, err := json.Marshal(entry)
	if err != nil {
		fmt.Fprintf(os.Stderr, "warning: failed to marshal log entry: %v\n", err)
		return
	}
	data = append(data, '\n')

	// Append to file. O_APPEND writes are atomic at the kernel level for
	// sizes under PIPE_BUF (typically 4KB), which JSONL entries always are.
	f, err := os.OpenFile(config.FilePath, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0644)
	if err != nil {
		fmt.Fprintf(os.Stderr, "warning: failed to open log file: %v\n", err)
		return
	}
	defer f.Close()

	if _, err := f.Write(data); err != nil {
		fmt.Fprintf(os.Stderr, "warning: failed to write log entry: %v\n", err)
	}
}

// rotateLog rotates the log file, keeping up to retainCount old files.
func rotateLog(config *LoggingConfig) {
	dir := filepath.Dir(config.FilePath)
	base := filepath.Base(config.FilePath)
	ext := filepath.Ext(base)
	name := strings.TrimSuffix(base, ext)

	// Find existing rotated files
	pattern := filepath.Join(dir, name+".*.jsonl")
	matches, _ := filepath.Glob(pattern)

	// Sort and remove excess
	if len(matches) >= config.RetainCount {
		sort.Strings(matches)
		for i := 0; i <= len(matches)-config.RetainCount; i++ {
			os.Remove(matches[i])
		}
	}

	// Rename current log
	ts := time.Now().UTC().Format("20060102T150405")
	rotated := filepath.Join(dir, fmt.Sprintf("%s.%s%s", name, ts, ext))
	os.Rename(config.FilePath, rotated)
}

// truncate shortens a string to maxLen characters.
func truncate(s string, maxLen int) string {
	if len(s) <= maxLen {
		return s
	}
	return s[:maxLen]
}

File: cli/embedded/sdks/typescript/logging.ts
import { join, dirname, basename } from "node:path";
import { mkdirSync, statSync, renameSync, readdirSync, unlinkSync, openSync, writeSync, closeSync, constants } from "node:fs";
import { homedir } from "node:os";
import type { SfaConfig } from "./config";

const DEFAULT_LOG_DIR = join(homedir(), ".local", "share", "single-file-agents", "logs");
const DEFAULT_LOG_FILE = join(DEFAULT_LOG_DIR, "executions.jsonl");
const DEFAULT_MAX_SIZE_BYTES = 50 * 1024 * 1024; // 50MB
const DEFAULT_RETAIN_COUNT = 5;

/**
 * JSONL log entry schema.
 */
export interface LogEntry {
  timestamp: string;
  agent: string;
  version: string;
  exitCode: number;
  durationMs: number;
  depth: number;
  callChain: string[];
  inputSummary: string;
  outputSummary: string;
  sessionId: string;
  meta?: Record<string, unknown>;
}

/**
 * Logging configuration resolved from env/config/defaults.
 */
export interface LoggingConfig {
  /** Resolved log file path */
  filePath: string;
  /** Whether logging is suppressed */
  suppressed: boolean;
  /** Max file size in bytes before rotation */
  maxSizeBytes: number;
  /** Number of rotated files to retain */
  retainCount: number;
}

/**
 * Resolve logging configuration from environment, config, and defaults.
 * Priority: SFA_LOG_FILE env → config logging.file → default path
 */
export function resolveLoggingConfig(config: SfaConfig, noLogFlag: boolean): LoggingConfig {
  const suppressed = noLogFlag || process.env.SFA_NO_LOG === "1";

  const filePath =
    process.env.SFA_LOG_FILE ??
    config.logging?.file ??
    DEFAULT_LOG_FILE;

  const maxSizeBytes = config.logging?.maxSize
    ? config.logging.maxSize * 1024 * 1024
    : DEFAULT_MAX_SIZE_BYTES;

  const retainCount = config.logging?.retainFiles ?? DEFAULT_RETAIN_COUNT;

  return { filePath, suppressed, maxSizeBytes, retainCount };
}

/**
 * Truncate a string to maxLen characters, appending "..." if truncated.
 */
function truncate(text: string, maxLen: number): string {
  if (text.length <= maxLen) return text;
  return text.slice(0, maxLen - 3) + "...";
}

/**
 * Create a log entry from execution data.
 */
export function createLogEntry(params: {
  agent: string;
  version: string;
  exitCode: number;
  startTime: number;
  depth: number;
  callChain: string[];
  sessionId: string;
  input: string;
  output: string;
  meta?: Record<string, unknown>;
}): LogEntry {
  return {
    timestamp: new Date().toISOString(),
    agent: params.agent,
    version: params.version,
    exitCode: params.exitCode,
    durationMs: Date.now() - params.startTime,
    depth: params.depth,
    callChain: params.callChain,
    inputSummary: truncate(params.input, 500),
    outputSummary: truncate(params.output, 500),
    sessionId: params.sessionId,
    ...(params.meta ? { meta: params.meta } : {}),
  };
}

/**
 * Ensure the log directory exists.
 */
function ensureLogDir(filePath: string): void {
  const dir = dirname(filePath);
  mkdirSync(dir, { recursive: true });
}

/**
 * Check if the log file needs rotation and rotate if so.
 */
function rotateIfNeeded(config: LoggingConfig): void {
  try {
    const stat = statSync(config.filePath);
    if (stat.size < config.maxSizeBytes) return;
  } catch {
    // File doesn't exist yet, no rotation needed
    return;
  }

  // Rotate: rename current file with timestamp suffix
  const now = new Date().toISOString().replace(/[:.]/g, "").slice(0, 15);
  const dir = dirname(config.filePath);
  const base = basename(config.filePath, ".jsonl");
  const rotatedName = `${base}-${now}.jsonl`;
  const rotatedPath = join(dir, rotatedName);

  try {
    renameSync(config.filePath, rotatedPath);
  } catch {
    // If rename fails, skip rotation
    return;
  }

  // Clean up old rotated files
  cleanupRotatedFiles(dir, base, config.retainCount);
}

/**
 * Remove rotated log files beyond the retain count.
 */
function cleanupRotatedFiles(dir: string, basePrefix: string, retainCount: number): void {
  try {
    const files = readdirSync(dir)
      .filter((f) => f.startsWith(basePrefix + "-") && f.endsWith(".jsonl"))
      .sort()
      .reverse();

    // Keep the most recent `retainCount` files, delete the rest
    for (let i = retainCount; i < files.length; i++) {
      try {
        unlinkSync(join(dir, files[i]));
      } catch {
        // Best effort
      }
    }
  } catch {
    // Best effort
  }
}

/**
 * Write a log entry to the JSONL file using O_APPEND for atomic writes.
 * This is best-effort: failures emit a warning to stderr but never affect exit code.
 */
export function writeLogEntry(entry: LogEntry, config: LoggingConfig): void {
  if (config.suppressed) return;

  try {
    ensureLogDir(config.filePath);
    rotateIfNeeded(config);

    const line = JSON.stringify(entry) + "\n";
    const fd = openSync(config.filePath, constants.O_WRONLY | constants.O_CREAT | constants.O_APPEND);
    try {
      writeSync(fd, line);
    } finally {
      closeSync(fd);
    }
  } catch (err: unknown) {
    // Non-blocking: warn on stderr, never fail
    process.stderr.write(
      `[sfa] warning: failed to write execution log: ${(err as Error).message}\n`,
    );
  }
}
